shared-module/displayio/Shape.c

```diff
@@ -66,8 +73,48 @@ void common_hal_displayio_shape_set_boundary(displayio_shape_t *self, uint16_t y
     if (self->mirror_x && (start_x > half_width || end_x > half_width)) {
         mp_raise_ValueError_varg(translate("Maximum x value when mirrored is %d"), half_width);
     }
+
+    uint16_t lower_x, upper_x;
+
+    // find x-boundaries for updating based on current data and start_x, end_x
+    lower_x = MIN(start_x, self->data[2 * y]);
+
+    if (self->mirror_x) {
+        upper_x = self->width-lower_x;
+    } else {
+        upper_x = 1 + MAX(end_x, self->data[2 * y + 1]);
+    }
+
     self->data[2 * y] = start_x;
     self->data[2 * y + 1] = end_x;
+
+    if (self->dirty_area.x1 == self->dirty_area.x2) { // Dirty region is empty
+        self->dirty_area.x1=lower_x;
+        self->dirty_area.x2=upper_x;
+        self->dirty_area.y1 = y;
+        if (self->mirror_y) {
+            self->dirty_area.y2 = self->height-y;
+        } else {
+            self->dirty_area.y2 = y+1;
+        }
```

Comment:
I think this is correct but I'd be simpler if you handled computing y bounds before saving it into dirty_area.

So, factor it out so you have lower_y and upper_y and then do the merge with the existing area.

Reply:
Ok, thanks for the additional prodding, I now see what you were getting at.  I cleaned up the y-portions similar to what I did with x, I hope it's clearer.

Also, I fixed the off-by-one error and realized that the dirty-rectangle tracking is "exclusive" for the `.x2` and `.y2` values, so I plastered that through the comments for posterity.  I think this is good now.